### PR TITLE
Fix update rate issues by working around MutliThreadedExecutor

### PR DESCRIPTION
**Pull request by Rayman**
Currently the MutliThreadedExecutor performance is very bad. This leads to controllers not meeting their update rate. This PR is a temporary workaround for these issues.

The current approach uses a `rclcpp` timer to execute the control loop. When used in combination with the `MutliThreadedExecutor`, the timers are not execute at their target frequency. I've converted the control loop to a while loop on a separate thread that uses `nanosleep` to execute the correct update rate. This means that `rclcpp` is not involved in the execution and leads to much better performance.

I've done some simple performance benchmarking. Using https://github.com/ros-controls/ros2_control_demos
with `update_rate` set to 500 Hz.

fix/update-rate-issues branch

```
$ ros2 topic hz /joint_states -w 500
average rate: 444.370
	min: 0.002s max: 0.003s std dev: 0.00016s window: 500
average rate: 443.508
	min: 0.002s max: 0.003s std dev: 0.00016s window: 500
average rate: 442.099
	min: 0.002s max: 0.003s std dev: 0.00016s window: 500
average rate: 445.781
	min: 0.002s max: 0.003s std dev: 0.00014s window: 500
average rate: 445.271
	min: 0.002s max: 0.003s std dev: 0.00016s window: 500
average rate: 443.078
	min: 0.002s max: 0.003s std dev: 0.00017s window: 500
average rate: 443.917
	min: 0.002s max: 0.003s std dev: 0.00015s window: 500
```

master branch
```
$ ros2 topic hz /joint_states -w 10
average rate: 4.811
	min: 0.067s max: 0.388s std dev: 0.12711s window: 7
average rate: 4.768
	min: 0.023s max: 0.388s std dev: 0.14797s window: 10
average rate: 11.099
	min: 0.002s max: 0.279s std dev: 0.08028s window: 10
average rate: 9.588
	min: 0.009s max: 0.277s std dev: 0.07749s window: 10
average rate: 7.070
	min: 0.009s max: 0.488s std dev: 0.13346s window: 10
average rate: 4.990
	min: 0.017s max: 0.712s std dev: 0.19685s window: 10
average rate: 4.472
	min: 0.002s max: 0.712s std dev: 0.23221s window: 10
```

I'm curious what you guys think about this workaround. Without this patch, I'm unable to run a control loop at 50Hz.

### CODE REVIEW

**Comment by bmagyar** [comment]
Could you please elaborate on the effects of this approach?

File: controller_manager/src/ros2_control_node.cpp [hunk 1]
```diff
@@ -32,24 +32,29 @@ int main(int argc, char ** argv)
...
-  auto timer = cm->create_wall_timer(
-    std::chrono::milliseconds(1000 / update_rate),
-    [&cm]() {
-      cm->read();
-      cm->update();
-      cm->write();
-    },
-    cm->deterministic_callback_group_);
+  // TODO: Due to issues with the MutliThreadedExecutor, this control loop does not rely on the executor. When the
+  // MutliThreadedExecutor issues are fixes, this loop should be converted back to a timer.
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:36]:
```suggestion
  // When the MutliThreadedExecutor issues are fixed (ros2/rclcpp#1168), this loop should be converted back to a timer.
```

File: controller_manager/src/ros2_control_node.cpp [hunk 2]
```diff
@@ -32,24 +32,29 @@ int main(int argc, char ** argv)
...
-
-  auto timer = cm->create_wall_timer(
-    std::chrono::milliseconds(1000 / update_rate),
-    [&cm]() {
-      cm->read();
-      cm->update();
-      cm->write();
-    },
-    cm->deterministic_callback_group_);
+  // TODO: Due to issues with the MutliThreadedExecutor, this control loop does not rely on the executor. When the
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:35]: This should restore the CI

```suggestion
  // TODO(anyone): Due to issues with the MutliThreadedExecutor, this control loop does not rely on the executor (see issue #260).
```

File: controller_manager/src/controller_manager.cpp [hunk 3]
```diff
@@ -92,8 +92,8 @@ ControllerManager::ControllerManager(
 
 void ControllerManager::init_services()
 {
-  deterministic_callback_group_ = create_callback_group(
-    rclcpp::CallbackGroupType::MutuallyExclusive);
+// deterministic_callback_group_ = create_callback_group(
```
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:95]:
```suggestion
// TODO(anyone): Due to issues with the MutliThreadedExecutor, this control loop does not rely on the executor (see issue #260).
// deterministic_callback_group_ = create_callback_group(
```

File: controller_manager/include/controller_manager/controller_manager.hpp [hunk 4]
```diff
@@ -126,7 +126,7 @@ class ControllerManager : public rclcpp::Node
    * Deterministic (real-time safe) callback group for the update function. Default behavior
    * is read hardware, update controller and finally write new values to the hardware.
    */
-  rclcpp::CallbackGroup::SharedPtr deterministic_callback_group_;
+  // rclcpp::CallbackGroup::SharedPtr deterministic_callback_group_;
```
Comment (destogl) [inline controller_manager/include/controller_manager/controller_manager.hpp:129]:
```suggestion
  // TODO(anyone): Due to issues with the MutliThreadedExecutor, this control loop does not rely on the executor (see issue #260).
  // rclcpp::CallbackGroup::SharedPtr deterministic_callback_group_;
```

**Review by destogl — CHANGES_REQUESTED** [review]
This is related to #260. It seems there is an issue in the depths of `rclcpp` (ros2/rclcpp#1487). There is also a solution that needs some more time... (ros2/rclcpp#1168).

The solution is IMHO legit until this is solved in `rclcpp`.

File: controller_manager/src/ros2_control_node.cpp (hunk 1 above)
Comment (Rayman) [reply controller_manager/src/ros2_control_node.cpp]: Done

File: controller_manager/src/ros2_control_node.cpp (hunk 2 above)
Comment (Rayman) [reply controller_manager/src/ros2_control_node.cpp]: Done

File: controller_manager/src/controller_manager.cpp (hunk 3 above)
Comment (Rayman) [reply controller_manager/src/controller_manager.cpp]: Done

File: controller_manager/include/controller_manager/controller_manager.hpp (hunk 4 above)
Comment (Rayman) [reply controller_manager/include/controller_manager/controller_manager.hpp:131]: Done

**Comment by Rayman** [comment]
@bmagyar I've added a simple benchmark to the description

**Review by bmagyar — APPROVED** [review]

**Review by Karsten1987 — APPROVED** [review]

**Comment by Karsten1987** [comment]
This PR actually broke the build on OSX.
```
/Users/karsten/workspace/ros2/ros2_control_ws/src/ros-controls/ros2_control/controller_manager/src/ros2_control_node.cpp:53:9: error: use of undeclared identifier 'clock_nanosleep'
        clock_nanosleep(CLOCK_MONOTONIC, 0, &duration, NULL);
``

**Comment by Rayman** [comment]
Sorry, I did not know OSX was a requirment. Should there be a CI for that?

Good suggestion to replace the call with `std::this_thread::sleep_for`. It should function the same and is more portable.

**Pull request merged**